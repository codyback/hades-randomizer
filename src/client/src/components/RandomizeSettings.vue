<template>
  <q-card class="settings q-ma-sm">
    <q-card-section>
      <div class="column">
        <q-toggle
          v-model="store.weightedRandomization"
          label="Weighted Randomization"
          color="primary"
        />
        <q-toggle
          v-model="store.hellMode"
          label="Hell Mode"
          color="secondary"
        />
      </div>
      <q-slider
        v-model="store.heatLevel"
        :min="minHeatLevel"
        :max="store.getHeatLevelMax"
        label
        label-always
        :label-value="`Heat Level: ${store.heatLevel}`"
        switch-label-side
        color="secondary"
      />
    </q-card-section>
    <q-card-section>
      <q-btn
        color="primary"
        label="Randomize"
        class="full-width"
        unelevated
      />
    </q-card-section>
  </q-card>
</template>

<script lang="ts" setup>
import { watch, ref } from 'vue';
import useStore from '../store';

const store = useStore();

const minHeatLevel = ref(1);

watch(() => store.hellMode, (hellMode) => {
  if (hellMode) {
    minHeatLevel.value = 5;
    store.hellModeHeats.forEach((heat) => {
      if (store.heatsFilter.includes(heat)) {
        store.heatsFilter.splice(store.heatsFilter.indexOf(heat), 1);
      }
    });
  } else {
    minHeatLevel.value = 1;
    if (!store.heatsFilter.includes('Personal Liability')) {
      store.heatsFilter.push('Personal Liability');
    }
  }
});

watch(() => store.getHeatLevelMax, (newValue) => {
  store.heatLevel = Math.min(store.heatLevel, newValue);
});
</script>

<style scoped lang="scss">
.settings {
  min-width: 30%;
}
</style>
