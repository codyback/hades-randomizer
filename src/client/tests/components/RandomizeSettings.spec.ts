import { describe, it, expect } from 'vitest';
import { shallowMount } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';
import useStore from '../../src/store';

import RandomizeSettings from '../../src/components/RandomizeSettings.vue';

const mountWith = () => {
  return shallowMount(RandomizeSettings, {
    global: {
      plugins: [createTestingPinia()],
    },
  });
};

describe('RandomizeSettings', () => {
  let wrapper: any;
  let store: any;

  beforeEach(() => {
    wrapper = mountWith();
    store = useStore();
  });

  it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('watchers', () => {
    describe('store.hellMode', () => {
      describe('is true', () => {
        beforeEach(() => {
          store.hellMode = false;
        });

        it('sets minHeatLevel to 5', async () => {
          expect(wrapper.vm.minHeatLevel).toBe(1);

          store.hellMode = true;
          await wrapper.vm.$nextTick();

          expect(wrapper.vm.minHeatLevel).toBe(5);
        });

        it('clears out hellModeHeats', async () => {
          expect(store.heatsFilter).toContain('Personal Liability');

          store.hellMode = true;
          await wrapper.vm.$nextTick();

          expect(store.heatsFilter).to.not.contain('Personal Liability');
        });
      });

      describe('is false', () => {
        beforeEach(() => {
          store.hellMode = true;
        });

        it('sets minHeatLevel to 1', async () => {
          expect(wrapper.vm.minHeatLevel).toBe(5);

          store.hellMode = false;
          await wrapper.vm.$nextTick();

          expect(wrapper.vm.minHeatLevel).toBe(1);
        });

        it('restores hellModeHeats', async () => {
          expect(store.heatsFilter).to.not.contain('Personal Liability');

          store.hellMode = false;
          await wrapper.vm.$nextTick();

          expect(store.heatsFilter).toContain('Personal Liability');
        });
      });
    });
  });
});
