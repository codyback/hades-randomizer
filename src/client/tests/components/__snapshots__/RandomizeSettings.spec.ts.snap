// Vitest Snapshot v1

exports[`RandomizeSettings > renders correctly 1`] = `
<q-card-stub
  bordered="false"
  class="settings q-ma-sm"
  data-v-c985db5f=""
  flat="false"
  square="false"
  tag="div"
/>
`;
